xmodule/capa/capa_problem.py

```diff
@@ -495,6 +503,41 @@ def get_grade_from_current_answers(self, student_answers):
 
         return newcmap
 
+    def get_grade_from_answers(self, student_answers, correct_map):
+        """
+        Gets the grade for the currently-saved problem state, but does not save it
+        to the block.
+
+        For new student_answers being graded, `student_answers` is a dict of all the
+        entries from request.POST, but with the first part of each key removed
+        (the string before the first "_").  Thus, for example,
+        input_ID123 -> ID123, and input_fromjs_ID123 -> fromjs_ID123.
+
+        For rescoring, `student_answers` is None.
+
+        Calls the Response for each question in this problem, to do the actual grading.
+        """
+        # old CorrectMap
+        oldcmap = correct_map
+
+        # start new with empty CorrectMap
+        newcmap = CorrectMap()
+        # Call each responsetype instance to do actual grading
+        for responder in self.responders.values():
+            # File objects are passed only if responsetype explicitly allows
+            # for file submissions.  But we have no way of knowing if
+            # student_answers contains a proper answer or the filename of
+            # an earlier submission, so for now skip these entirely.
+            # TODO: figure out where to get file submissions when rescoring.
```

Comment:
is this still pending?

Reply:
This new method is based on [this method]( https://github.com/openedx/edx-platform/blob/dfe9032dd900b100b8405f28d34f9f387ca5a9e3/xmodule/capa/capa_problem.py#L466-L504)